### PR TITLE
updated celery cronjob with improvements.

### PR DESCRIPTION
**Pull request by prem-fissionhq**
##### ENVIRONMENTS AFFECTED

Old PR: https://github.com/dimagi/commcare-cloud/pull/4206

Changes:
1. added two new variables to control hours for the soft and hard kill of celery tasks.
2. added a new cronjob script. The script will not throw error message if no process was found.

Please review and let me know your suggestions. thanks

### CODE REVIEW

File: src/commcare_cloud/ansible/roles/commcarehq/templates/celery_task_kill.sh.j2 [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+HOURS=$1
+MAX_MINUTES=`expr ${HOURS} \* 60 \* 60`
+
+SIG_NAME=$2
+
+PID=$(ps -eo comm,pid,etimes | awk '/celeryd:/ { if ( $4 > '"${MAX_MINUTES}"' ) { print $3 } }');
+
+if [ "$PID" != "" ];
+then
+	kill --verbose -"${SIG_NAME}" $(ps -eo comm,pid,etimes | awk '/celeryd:/ { if ( $4 > '"${MAX_MINUTES}"' ) { print $3 } }');
```
Comment (snopoke) [inline src/commcare_cloud/ansible/roles/commcarehq/templates/celery_task_kill.sh.j2:15]: I don't think `kill` has a verbose option.

Perhaps you could do something like:
1. get list of program names + PIDs to kill
2. convert that to list of PIDs
3. Echo list of program names + PIDs
4. issue kill command
5. Check that all PIDs are terminated (? maybe)

File: src/commcare_cloud/ansible/roles/commcarehq/templates/celery_task_kill.sh.j2 [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+else
+	echo "No Process to kill";
```
Comment (snopoke) [inline src/commcare_cloud/ansible/roles/commcarehq/templates/celery_task_kill.sh.j2:17]: Won't this output also trigger an email?

File: src/commcare_cloud/ansible/roles/commcarehq/templates/celery_task_kill.sh.j2 [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+# Process to soft and hard kill old celery jobs.
+#
+
+HOURS=$1
+MAX_MINUTES=`expr ${HOURS} \* 60 \* 60`
+
+SIG_NAME=$2
+
+PID=$(ps -eo comm,pid,etimes | awk '/celeryd:/ { if ( $4 > '"${MAX_MINUTES}"' ) { print $3 } }');
```
Comment (snopoke) [inline src/commcare_cloud/ansible/roles/commcarehq/templates/celery_task_kill.sh.j2:11]: I think $4 & $3 are incorrect (should be $3 and $2), also not 100% sure about the quoting around MAX_MINUTES

File: src/commcare_cloud/ansible/group_vars/all.yml [hunk 4]
```diff
@@ -155,3 +155,8 @@ keepalived_recepient_email: '{{ root_email }}'
 # this parameter is to set backups hour frequency, by default it will take backups for every '3' hours 
 # if hourly backups are enabled. 
 nadir_hour_frequency: 3
+
+# celery variables to soft/hard kill tasks running for long hours.
+# To enable cronjob give hours hard_kill: eg: 46, eg: soft_kill: 36. 'False' will remove cron jobs.
+celery_hours_before_hard_kill: False
+celery_hours_before_soft_kill: False
```
[the remark this answers is not in the document]
Comment (prem-fissionhq) [reply src/commcare_cloud/ansible/group_vars/all.yml:162]: I tried with None and some reason the absent case was not working. So I had to make it false.

**Comment by prem-fissionhq** [comment]
@snopoke I will offline with you on this items. thanks

File: src/commcare_cloud/ansible/roles/commcarehq/templates/celery_task_kill.sh.j2 (hunk 3 above)
Comment (prem-fissionhq) [reply src/commcare_cloud/ansible/roles/commcarehq/templates/celery_task_kill.sh.j2:11]: MAX_Minutes is working as expected. Tested.

File: src/commcare_cloud/ansible/roles/commcarehq/templates/celery_task_kill.sh.j2 (hunk 1 above)
Comment (snopoke) [reply src/commcare_cloud/ansible/roles/commcarehq/templates/celery_task_kill.sh.j2:15]: OK interesting. The ubuntu man page doesn't indicate a verbose option: https://manpages.ubuntu.com/manpages/bionic/man1/kill.1.html

When you tested it were the processes that matched? Perhpas it didn't execute the kill command.

File: src/commcare_cloud/ansible/roles/commcarehq/templates/celery_task_kill.sh.j2 (hunk 3 above)
Comment (snopoke) [reply src/commcare_cloud/ansible/roles/commcarehq/templates/celery_task_kill.sh.j2:11]: I think if you test it out you'll find it's not working. The `ps` command only has 3 columns so $4 will be blank. $3 is the elapsed time and $2 is the PID:
```
~$ ps -eo comm,pid,etimes | head | ps -eo comm,pid,etimes | head | awk '{ print $2, $3 }'

PID ELAPSED
1 1372223
2 1372223
3 1372223
4 1372223
6 1372223
9 1372223
10 1372223
11 1372223
12 1372223
```

File: src/commcare_cloud/ansible/deploy_commcarehq.yml [hunk 5]
```diff
@@ -138,4 +138,4 @@
   hosts: celery
   tasks:
     - include_tasks: roles/commcarehq/tasks/celery_cron.yml
```
Comment (dannyroberts) [inline src/commcare_cloud/ansible/deploy_commcarehq.yml:140]: I tried to run this on india today with

```
cchq india deploy-stack --limit celery --tags cron
```

In order for this task to get picked up I had to change it to `import_tasks`. https://docs.ansible.com/ansible/latest/user_guide/playbooks_reuse.html for difference between import_* and include_*.

I'm not sure if that's the right solution, but without it I think the `cron` tags in `celery_cron.yml` might not do anything.

File: src/commcare_cloud/ansible/roles/commcarehq/templates/celery_task_kill.sh.j2 (hunk 3 above)
Comment (prem-fissionhq) [reply src/commcare_cloud/ansible/roles/commcarehq/templates/celery_task_kill.sh.j2:11]: @snopoke The command which I'm using is slightly different. The celeryd process has some spaces in the output therefore it giving me four slots. 
example1:
```
ps -eo comm,pid,etimes | grep celeryd
[celeryd: celer  1167      62
[celeryd: celer  1210      43
[celeryd: celer  1212      43
[celeryd: celer  1254      33
[celeryd: celer  1258      23
[celeryd: celer  1259      22
```
example2: 
```
ps -eo comm,pid,etimes | awk '/celeryd:/'
[celeryd: celer  1167      53
[celeryd: celer  1210      34
[celeryd: celer  1212      34
[celeryd: celer  1253      24
[celeryd: celer  1254      24
```
actual cmd with test minutes 12960: 
```
---  command has space [celeryd: celer ---
ps -eo comm,pid,etimes | awk '/celeryd:/ { if ( $4 > 12960 ) { print $1, $2, $3, $4 } }';
command, pid, etimes
[celeryd: celer 11972 126769
[celeryd: celer 13088 126565
[celeryd: celer 28246 35571
[celeryd: celer 28247 35571
[celeryd: celer 28249 35571
``` 
pid result: took from above " [celeryd: celer **28249** 35571 "
```
> ps -ef | grep 28259
cchq      1526 28259  0 05:55 ?        00:00:00 [celeryd: celery@<privateip>_saved_exports_queue_0.1602100766_timestamp:ForkPoolWorker-595]
pkasha    1610   840  0 05:56 pts/0    00:00:00 grep --color=auto 28259
cchq     28259 28241  1 Oct07 ?        00:09:40 [celeryd: celery@privateip_saved_exports_queue_0.1602100766_timestamp:MainProcess] -active- (-A corehq worker --hostname=privateip_saved_exports_queue_0.1602100766_timestamp --events --loglevel=INFO --without-gossip --queues=saved_exports_queue --autoscale=3,0 -Ofair --prefetch-multiplier=4)
```
```
pid result: took from above " [celeryd: celer 28249 **35571** "
> ps -ef | grep 35571
pkasha    1912   840  0 05:58 pts/0    00:00:00 grep --color=auto 35571
```

File: src/commcare_cloud/ansible/roles/commcarehq/templates/celery_task_kill.sh.j2 (hunk 1 above)
Comment (prem-fissionhq) [reply src/commcare_cloud/ansible/roles/commcarehq/templates/celery_task_kill.sh.j2:15]: yes, the --verbose could have failed and I should have tested/caught it, I moved forward with it as it was mentioned in man-pages document. Going forward I will cross check twice. I removed it from the file now. thanks Simon.

File: src/commcare_cloud/ansible/deploy_commcarehq.yml (hunk 5 above)
Comment (prem-fissionhq) [reply src/commcare_cloud/ansible/deploy_commcarehq.yml:140]: @dannyroberts sure, I will make a note of it. thanks

File: src/commcare_cloud/ansible/roles/commcarehq/templates/celery_task_kill.sh.j2 (hunk 3 above)
Comment (snopoke) [reply src/commcare_cloud/ansible/roles/commcarehq/templates/celery_task_kill.sh.j2:11]: :+1: thanks for the explanation

File: src/commcare_cloud/ansible/deploy_commcarehq.yml (hunk 5 above)
Comment (snopoke) [reply src/commcare_cloud/ansible/deploy_commcarehq.yml:140]: I think it is generally preferable to use `import_tasks`

File: src/commcare_cloud/ansible/group_vars/all.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (snopoke) [reply src/commcare_cloud/ansible/group_vars/all.yml:162]: See https://github.com/dimagi/commcare-cloud/pull/4245/files#r500412967

[the remark this answers is not in the document]
Comment (prem-fissionhq) [reply src/commcare_cloud/ansible/group_vars/all.yml]: I tested myself and fixed it now. Thanks @snopoke

File: src/commcare_cloud/ansible/roles/commcarehq/tasks/celery_cron.yml [hunk 6]
```diff
@@ -10,6 +11,7 @@
     weekday: "0,1,2,3,4,5,6"
     user: root
     cron_file: celery_sigterm
+    state: absent
```
[the remark this answers is not in the document]
Comment (prem-fissionhq) [reply src/commcare_cloud/ansible/roles/commcarehq/tasks/celery_cron.yml:9]: @dannyroberts done. updated file with changes.

File: src/commcare_cloud/ansible/roles/commcarehq/tasks/celery_cron.yml [hunk 7]
```diff
@@ -23,5 +25,45 @@
...
+- name: Celery cron jobs to soft kill old tasks
+  become: true
+  cron:
+    name: "Soft kill old celery processes"
+    job: "/usr/local/sbin/celery_task_kill.sh {{ celery_hours_before_soft_kill }} SIGTERM "
+    minute: 15
+    hour: "*"
+    user: root
+    cron_file: celery_sigterm
+    state: "{{ 'present' if celery_hours_before_soft_kill != False else 'absent' }}"
```
[the remark this answers is not in the document]
Comment (prem-fissionhq) [reply src/commcare_cloud/ansible/roles/commcarehq/tasks/celery_cron.yml]: @dannyroberts Updated file with changes and working as expected.

File: src/commcare_cloud/ansible/roles/commcarehq/tasks/celery_cron.yml [hunk 8]
```diff
@@ -1,27 +1,59 @@
...
+- name: Copy celery cron script
+  become: yes
+  template:
+    src: "celery_task_kill.sh.j2"
+    dest: "/usr/local/sbin/celery_task_kill.sh"
+    group: root
+    owner: root
+    mode: 0744
+    backup: yes
+  when: "celery_hours_before_soft_kill and celery_hours_before_hard_kill"
```
Comment (snopoke) [inline src/commcare_cloud/ansible/roles/commcarehq/tasks/celery_cron.yml:31]: The tasks below only check one of the vars so the script needs to be present if either is set
```suggestion
  when: "celery_hours_before_soft_kill or celery_hours_before_hard_kill"
```

File: src/commcare_cloud/ansible/roles/commcarehq/tasks/celery_cron.yml [hunk 9]
```diff
@@ -23,5 +25,45 @@
...
+- name: Celery cron job to kill old tasks
+  become: true
+  cron:
+    name: "Hard kill old celery processes"
+    job: "/usr/local/sbin/celery_task_kill.sh {{ celery_hours_before_hard_kill }} SIGKILL "
+    minute: 0
+    hour: "*"
+    user: root
+    cron_file: celery_sigkill
+    state: "{{ 'present' if celery_hours_before_hard_kill != False else 'absent' }}"
```
[the remark this answers is not in the document]
Comment (prem-fissionhq) [reply src/commcare_cloud/ansible/roles/commcarehq/tasks/celery_cron.yml]: Suggested changes have been correct in the file.

File: src/commcare_cloud/ansible/deploy_commcarehq.yml (hunk 5 above)
Comment (prem-fissionhq) [reply src/commcare_cloud/ansible/deploy_commcarehq.yml]: @dannyroberts updated files as you suggested. please review and let me know if any changes are necessary.

**Review by snopoke — COMMENTED** [review]
This all looks good to me. One thing I'm still not clear on is how any notification will happen if processes are killed.

**Comment by prem-fissionhq** [comment]
> This all looks good to me. One thing I'm still not clear on is how any notification will happen if processes are killed.

@snopoke as of now, we are not sending any emails/notifications with the details of the celery tasks that were killed.

File: src/commcare_cloud/ansible/roles/commcarehq/templates/celery_task_kill.sh.j2 [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+MAX_MINUTES=`expr ${HOURS} \* 60 \* 60`
+
+SIG_NAME=$2
+
+PIDS=$(ps -eo comm,pid,etimes | awk '/celeryd:/ { if ( $4 > '"${MAX_MINUTES}"' ) { print $3 } }');
+
+if [ "$PIDS" != "" ]; then
+  for pid in ${PIDS}
+  do
+     #echo "kill -${SIG_NAME} ${pid}"
```
Comment (dannyroberts) [inline src/commcare_cloud/ansible/roles/commcarehq/templates/celery_task_kill.sh.j2:16]: If you uncomment this, would it send us an email? I think that might be preferable if it does

Comment (prem-fissionhq) [reply src/commcare_cloud/ansible/roles/commcarehq/templates/celery_task_kill.sh.j2:16]: @dannyroberts, not completely sure. I have enabled it in the staging celery server will monitor for sometime and let you know. thanks

Comment (snopoke) [reply src/commcare_cloud/ansible/roles/commcarehq/templates/celery_task_kill.sh.j2:16]: Getting this in emails from staging:
```
kill -SIGTERM 8449
kill -SIGTERM 8453
kill -SIGTERM 9798
```

Would be great to get more detail:
```
if [ "$PIDS" != "" ]; then
  echo "Killing stale celery workers:
  echo "ETIME COMMAND
  for pid in ${PIDS}
  do
     echo `ps -o etime= -o args= $PID`
     ....
```

**Review by snopoke — APPROVED** [review]
I'm going to approve these as is so that we can get them rolled out ASAP to stop the flood of cron emails.

I think the current state of the PR is better than what is deployed.

File: src/commcare_cloud/ansible/roles/commcarehq/templates/celery_task_kill.sh.j2 (hunk 10 above)
Comment (prem-fissionhq) [reply src/commcare_cloud/ansible/roles/commcarehq/templates/celery_task_kill.sh.j2:16]: @snopoke thanks for the suggestions. Will update the script and commit the changes.

Comment (prem-fissionhq) [reply src/commcare_cloud/ansible/roles/commcarehq/templates/celery_task_kill.sh.j2]: @snopoke @dannyroberts updated the template like this
```
if [ "$PIDS" != "" ]; then
  echo "Killing stale celery workers:"
  echo "Process information with start and elapsed time:"
  for pid in ${PIDS}
  do
     echo "`ps -p ${pid} -o pid,start,etime`"
     echo "kill -${SIG_NAME} ${pid}"
     kill -${SIG_NAME} ${pid}
  done
fi
```
sample output:
```
Killing stale celery workers:
Process information with start and elapsed time:
  PID  STARTED     ELAPSED
20852 14:06:37    01:31:25
kill -SIGTERM 20852
  PID  STARTED     ELAPSED
20854 14:06:38    01:31:24
kill -SIGTERM 20854
  PID  STARTED     ELAPSED
20856 14:06:38    01:31:24
kill -SIGTERM 20856
```

File: src/commcare_cloud/ansible/roles/commcarehq/templates/celery_task_kill.sh.j2 [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+
+PIDS=$(ps -eo comm,pid,etimes | awk '/celeryd:/ { if ( $4 > '"${MAX_MINUTES}"' ) { print $3 } }');
+
+if [ "$PIDS" != "" ]; then
+  echo "Killing stale celery workers:"
+  echo "Process information with start and elapsed time:"
+  for pid in ${PIDS}
+  do
+     echo "`ps -p ${pid} -o pid,start,etime,cmd`"
+     echo "kill -${SIG_NAME} ${pid}"
```
Comment (snopoke) [inline src/commcare_cloud/ansible/roles/commcarehq/templates/celery_task_kill.sh.j2:19]: you can remove this line now
```suggestion
```

**Review by snopoke — APPROVED** [review]

File: src/commcare_cloud/ansible/group_vars/all.yml [hunk 12]
```diff
@@ -156,6 +156,11 @@ keepalived_recepient_email: '{{ root_email }}'
 # if hourly backups are enabled. 
 nadir_hour_frequency: 3
 
+# celery variables to soft/hard kill tasks running for long hours.
+# To enable cronjob give hours hard_kill: eg: 46, eg: soft_kill: 36. 'False' will remove cron jobs.
```
Comment (dannyroberts) [inline src/commcare_cloud/ansible/group_vars/all.yml:160]:
```suggestion
# To enable cron job to hard/soft kill celery processes set these thresholds. Use null/~ to disable cron job.
```

**Review by dannyroberts — APPROVED** [review]
Looks good. Good to merge now or after committing any final PR feedback

**Comment by prem-fissionhq** [comment]
@dannyroberts @snopoke Applied the changes. thanks

File: src/commcare_cloud/ansible/roles/commcarehq/tasks/celery_cron.yml [hunk 13]
```diff
@@ -1,27 +1,59 @@
...
+    name: "Sending SIGTERM to celery processes running morethan 46h"
+    cron_file: celery_sigkill
+    state: absent
+  tags:
+    - cron
+
+- name: Copy celery cron script
+  become: yes
+  template:
+    src: "celery_task_kill.sh.j2"
```
Comment (dannyroberts) [inline src/commcare_cloud/ansible/roles/commcarehq/tasks/celery_cron.yml:25]: For some reason this line is giving me

```
fatal: [10.202.12.80]: FAILED! => {"changed": false, "msg": "Could not find or access 'celery_task_kill.sh.j2'\nSearched in:\n\t/Users/danielroberts/dimagi/commcare-cloud/src/commcare_cloud/ansible/roles/commcarehq/tasks/templates/celery_task_kill.sh.j2\n\t/Users/danielroberts/dimagi/commcare-cloud/src/commcare_cloud/ansible/roles/commcarehq/tasks/celery_task_kill.sh.j2\n\t/Users/danielroberts/dimagi/commcare-cloud/src/commcare_cloud/ansible/templates/celery_task_kill.sh.j2\n\t/Users/danielroberts/dimagi/commcare-cloud/src/commcare_cloud/ansible/celery_task_kill.sh.j2 on the Ansible Controller.\nIf you are using a module and expect the file to exist on the remote, see the remote_src option"}
```
Here's the text there better formatted:
```
Could not find or access 'celery_task_kill.sh.j2'
Searched in:
	/Users/danielroberts/dimagi/commcare-cloud/src/commcare_cloud/ansible/roles/commcarehq/tasks/templates/celery_task_kill.sh.j2
	/Users/danielroberts/dimagi/commcare-cloud/src/commcare_cloud/ansible/roles/commcarehq/tasks/celery_task_kill.sh.j2
	/Users/danielroberts/dimagi/commcare-cloud/src/commcare_cloud/ansible/templates/celery_task_kill.sh.j2
	/Users/danielroberts/dimagi/commcare-cloud/src/commcare_cloud/ansible/celery_task_kill.sh.j2 on the Ansible Controller.
If you are using a module and expect the file to exist on the remote, see the remote_src option
```
If I change this line to `src: "../templates/celery_task_kill.sh.j2"` instead, then it runs fine. Not sure if it is something wrong with my setup or with this line. I would expect it to work as it is here just fine, just like https://github.com/dimagi/commcare-cloud/blob/673a27383c7371da5012cd3a6fb76689b7176e6f/src/commcare_cloud/ansible/roles/nginx/tasks/site.yml#L24 does.

I ran this this way: `cchq production deploy-stack --limit celery --tags cron`

Comment (snopoke) [reply src/commcare_cloud/ansible/roles/commcarehq/tasks/celery_cron.yml:25]: @dannyroberts that's odd. It appears to have looked in the correct place as well: `/Users/danielroberts/dimagi/commcare-cloud/src/commcare_cloud/ansible/roles/commcarehq/tasks/templates/celery_task_kill.sh.j2`. Did that file not exist?

Comment (snopoke) [reply src/commcare_cloud/ansible/roles/commcarehq/tasks/celery_cron.yml:25]: Is this in `deploy_commcarehq.yml`? that would make total sense. Why are we even doing it like this at all and not just running the tasks as part of the main role exection?

Comment (dannyroberts) [reply src/commcare_cloud/ansible/roles/commcarehq/tasks/celery_cron.yml:25]: That's a good point. It's nice to have these related tasks be in their own file, but I agree it'd be cleaner to import these tasks from the `celery.yml` tasks file (also in the `commcarehq` role) which is already executed from `deploy_commcarehq.yml` in the same way

Comment (dannyroberts) [reply src/commcare_cloud/ansible/roles/commcarehq/tasks/celery_cron.yml:25]: Looks like we never got around to this. @prem-fissionhq, @sanjay2916 bumped into this today and may make similar changes locally and/or PR the fix, so just make sure you coordinate with him on it

Comment (prem-fissionhq) [reply src/commcare_cloud/ansible/roles/commcarehq/tasks/celery_cron.yml:25]: @dannyroberts I'm working on it. thanks

Comment (prem-fissionhq) [reply src/commcare_cloud/ansible/roles/commcarehq/tasks/celery_cron.yml:25]: @dannyroberts this worked for me too. 
"src: "../templates/celery_task_kill.sh.j2" without any issue, testing other ways too.

Comment (dannyroberts) [reply src/commcare_cloud/ansible/roles/commcarehq/tasks/celery_cron.yml:25]: While that does work, I think what I suggested above is the way that ansible would recommend. By using `import_role` a number of default paths are set up, including the template directory, letting you use the normal conventions.

Comment (prem-fissionhq) [reply src/commcare_cloud/ansible/roles/commcarehq/tasks/celery_cron.yml:25]: @dannyroberts updated the file as per your suggestions and tested in staging and it worked fine for us. 
please review, thanks
https://github.com/dimagi/commcare-cloud/pull/4329

**Pull request merged**